File: src/ai-node/lifecycle/nodeLifecycle.js
export const NODE_LIFECYCLE_STATES = Object.freeze({
  UNCONFIGURED: "unconfigured",
  BOOTSTRAP_CONNECTING: "bootstrap_connecting",
  BOOTSTRAP_CONNECTED: "bootstrap_connected",
  CORE_DISCOVERED: "core_discovered",
  REGISTRATION_PENDING: "registration_pending",
  PENDING_APPROVAL: "pending_approval",
  TRUSTED: "trusted",
  CAPABILITY_SETUP_PENDING: "capability_setup_pending",
  OPERATIONAL: "operational",
  DEGRADED: "degraded",
});

const ALLOWED_TRANSITIONS = new Map([
  [NODE_LIFECYCLE_STATES.UNCONFIGURED, new Set([NODE_LIFECYCLE_STATES.BOOTSTRAP_CONNECTING])],
  [NODE_LIFECYCLE_STATES.BOOTSTRAP_CONNECTING, new Set([NODE_LIFECYCLE_STATES.BOOTSTRAP_CONNECTED])],
  [NODE_LIFECYCLE_STATES.BOOTSTRAP_CONNECTED, new Set([NODE_LIFECYCLE_STATES.CORE_DISCOVERED])],
  [NODE_LIFECYCLE_STATES.CORE_DISCOVERED, new Set([NODE_LIFECYCLE_STATES.REGISTRATION_PENDING])],
  [NODE_LIFECYCLE_STATES.REGISTRATION_PENDING, new Set([NODE_LIFECYCLE_STATES.PENDING_APPROVAL])],
  [NODE_LIFECYCLE_STATES.PENDING_APPROVAL, new Set([NODE_LIFECYCLE_STATES.TRUSTED])],
  [NODE_LIFECYCLE_STATES.TRUSTED, new Set([NODE_LIFECYCLE_STATES.CAPABILITY_SETUP_PENDING])],
  [NODE_LIFECYCLE_STATES.CAPABILITY_SETUP_PENDING, new Set([NODE_LIFECYCLE_STATES.OPERATIONAL])],
  [NODE_LIFECYCLE_STATES.OPERATIONAL, new Set([])],
  [NODE_LIFECYCLE_STATES.DEGRADED, new Set([NODE_LIFECYCLE_STATES.OPERATIONAL])],
]);

const STATE_VALUES = new Set(Object.values(NODE_LIFECYCLE_STATES));
const DEGRADABLE_STATES = new Set([
  NODE_LIFECYCLE_STATES.BOOTSTRAP_CONNECTING,
  NODE_LIFECYCLE_STATES.BOOTSTRAP_CONNECTED,
  NODE_LIFECYCLE_STATES.CORE_DISCOVERED,
  NODE_LIFECYCLE_STATES.REGISTRATION_PENDING,
  NODE_LIFECYCLE_STATES.PENDING_APPROVAL,
  NODE_LIFECYCLE_STATES.TRUSTED,
  NODE_LIFECYCLE_STATES.CAPABILITY_SETUP_PENDING,
  NODE_LIFECYCLE_STATES.OPERATIONAL,
]);

function canTransition(from, to) {
  if (to === NODE_LIFECYCLE_STATES.DEGRADED && DEGRADABLE_STATES.has(from)) {
    return true;
  }
  return ALLOWED_TRANSITIONS.get(from)?.has(to) === true;
}

export function createNodeLifecycle(options = {}) {
  const logger = options.logger ?? console;
  const onTransition = options.onTransition ?? (() => {});
  let state = NODE_LIFECYCLE_STATES.UNCONFIGURED;

  return {
    getState() {
      return state;
    },
    canTransitionTo(nextState) {
      return canTransition(state, nextState);
    },
    transitionTo(nextState, meta = {}) {
      if (!STATE_VALUES.has(nextState)) {
        throw new Error(`unknown lifecycle state: ${nextState}`);
      }
      if (!canTransition(state, nextState)) {
        throw new Error(`invalid state transition: ${state} -> ${nextState}`);
      }

      const previous = state;
      state = nextState;
      logger.info?.("[state-transition]", { from: previous, to: nextState, ...meta });
      onTransition({ from: previous, to: nextState, meta });
      return state;
    },
  };
}

File: src/ai_node/lifecycle/node_lifecycle.py
from enum import Enum
from typing import Callable, Dict, Optional, Set

from ai_node.diagnostics.onboarding_logger import OnboardingDiagnosticsLogger


class NodeLifecycleState(str, Enum):
    UNCONFIGURED = "unconfigured"
    BOOTSTRAP_CONNECTING = "bootstrap_connecting"
    BOOTSTRAP_CONNECTED = "bootstrap_connected"
    CORE_DISCOVERED = "core_discovered"
    REGISTRATION_PENDING = "registration_pending"
    PENDING_APPROVAL = "pending_approval"
    TRUSTED = "trusted"
    CAPABILITY_SETUP_PENDING = "capability_setup_pending"
    OPERATIONAL = "operational"
    DEGRADED = "degraded"


ALLOWED_TRANSITIONS: Dict[NodeLifecycleState, Set[NodeLifecycleState]] = {
    NodeLifecycleState.UNCONFIGURED: {
        NodeLifecycleState.BOOTSTRAP_CONNECTING,
        NodeLifecycleState.TRUSTED,
    },
    NodeLifecycleState.BOOTSTRAP_CONNECTING: {
        NodeLifecycleState.BOOTSTRAP_CONNECTED,
        NodeLifecycleState.UNCONFIGURED,
    },
    NodeLifecycleState.BOOTSTRAP_CONNECTED: {NodeLifecycleState.CORE_DISCOVERED},
    NodeLifecycleState.CORE_DISCOVERED: {NodeLifecycleState.REGISTRATION_PENDING},
    NodeLifecycleState.REGISTRATION_PENDING: {NodeLifecycleState.PENDING_APPROVAL},
    NodeLifecycleState.PENDING_APPROVAL: {NodeLifecycleState.TRUSTED},
    NodeLifecycleState.TRUSTED: {NodeLifecycleState.CAPABILITY_SETUP_PENDING},
    NodeLifecycleState.CAPABILITY_SETUP_PENDING: {NodeLifecycleState.OPERATIONAL},
    NodeLifecycleState.OPERATIONAL: set(),
    NodeLifecycleState.DEGRADED: {NodeLifecycleState.OPERATIONAL},
}

DEGRADABLE_STATES: Set[NodeLifecycleState] = {
    NodeLifecycleState.BOOTSTRAP_CONNECTING,
    NodeLifecycleState.BOOTSTRAP_CONNECTED,
    NodeLifecycleState.CORE_DISCOVERED,
    NodeLifecycleState.REGISTRATION_PENDING,
    NodeLifecycleState.PENDING_APPROVAL,
    NodeLifecycleState.TRUSTED,
    NodeLifecycleState.CAPABILITY_SETUP_PENDING,
    NodeLifecycleState.OPERATIONAL,
}


class NodeLifecycle:
    def __init__(
        self,
        logger,
        on_transition: Optional[Callable[[dict], None]] = None,
    ) -> None:
        self._state = NodeLifecycleState.UNCONFIGURED
        self._logger = logger
        self._diag = OnboardingDiagnosticsLogger(logger)
        self._on_transition = on_transition or (lambda _: None)

    def get_state(self) -> NodeLifecycleState:
        return self._state

    def can_transition_to(self, next_state: NodeLifecycleState) -> bool:
        return _can_transition(self._state, next_state)

    def transition_to(self, next_state: NodeLifecycleState, meta: Optional[dict] = None) -> NodeLifecycleState:
        if not isinstance(next_state, NodeLifecycleState):
            raise ValueError(f"unknown lifecycle state: {next_state}")
        if not _can_transition(self._state, next_state):
            raise ValueError(f"invalid state transition: {self._state.value} -> {next_state.value}")

        previous = self._state
        self._state = next_state
        payload = {"from": previous.value, "to": next_state.value, **(meta or {})}
        if hasattr(self._logger, "info"):
            self._logger.info("[state-transition] %s", payload)
        self._diag.state_transition(payload)
        self._on_transition({"from": previous, "to": next_state, "meta": meta or {}})
        return self._state

    def reset_to_unconfigured(self, meta: Optional[dict] = None) -> NodeLifecycleState:
        previous = self._state
        self._state = NodeLifecycleState.UNCONFIGURED
        payload = {"from": previous.value, "to": NodeLifecycleState.UNCONFIGURED.value, **(meta or {})}
        if hasattr(self._logger, "info"):
            self._logger.info("[state-transition] %s", payload)
        self._diag.state_transition(payload)
        self._on_transition(
            {
                "from": previous,
                "to": NodeLifecycleState.UNCONFIGURED,
                "meta": meta or {},
            }
        )
        return self._state


def _can_transition(from_state: NodeLifecycleState, to_state: NodeLifecycleState) -> bool:
    if to_state == NodeLifecycleState.DEGRADED and from_state in DEGRADABLE_STATES:
        return True
    return to_state in ALLOWED_TRANSITIONS.get(from_state, set())
